File: task16/main.cpp
#include <iostream>
#include <vector>
#include <string>
// Вариант с файлом:
// #include <fstream>
using namespace std;

vector<int> bmSearch(string& text, string& pattern) {
    int n = text.size();
    int m = pattern.size();
    vector<int> results;

    vector<int> d(256, m);
    for (int i = 0; i < m; i++)
        d[(unsigned char)pattern[i]] = m - i - 1;

    int i = m - 1;
    while (i < n) {
        int j = m - 1;
        int k = i;
        while (j >= 0 && pattern[j] == text[k]) {
            j--;
            k--;
        }
        if (j == -1) {
            results.push_back(i - m + 1);
            i++;
        } else {
            i += d[(unsigned char)text[i]];
        }
    }

    return results;
}

int main() {
    // Вариант с файлом:
    // ifstream f("input.txt");
    // string text, pattern;
    // getline(f, text);
    // getline(f, pattern);

    string text, pattern;
    cout << "Введите текст: ";
    getline(cin, text);
    cout << "Введите образец: ";
    getline(cin, pattern);

    vector<int> positions = bmSearch(text, pattern);

    cout << "Вхождения на позициях: ";
    for (int pos : positions)
        cout << pos << " ";

    return 0;
}

File: task16/main.py
# Вариант с файлом:
# with open("input.txt") as f:
#     lines = f.read().splitlines()
#     text = lines[0]
#     pattern = lines[1]

def bm_search(text, pattern):
    n = len(text)
    m = len(pattern)
    results = []

    d = [m] * 256
    for i in range(m):
        d[ord(pattern[i])] = m - i - 1

    i = m - 1
    while i < n:
        j = m - 1
        k = i
        while j >= 0 and pattern[j] == text[k]:
            j -= 1
            k -= 1
        if j == -1:
            results.append(i - m + 1)
            i += 1
        else:
            i += d[ord(text[i])]

    return results

text = input("Введите текст: ")
pattern = input("Введите образец: ")
print("Вхождения на позициях:", bm_search(text, pattern))